### PR TITLE
feat: Ability to provide custom onBlur and onChange handlers

### PR DESCRIPTION
**Pull request by michalbundyra**
Related to #4977 and #4751

If we want to provide custom `onBlur` or `onChange` we can just do:

```jsx
<input type="text" {...register('test', {
    onChange: (e) => {
        // custom onChange handler
    },
    onBlur: (e) => {
        // custom onBlur handler
    },
})}/>
```

### CODE REVIEW

File: src/__tests__/useForm/register.test.tsx [hunk 1]
```diff
@@ -1014,4 +1014,68 @@ describe('register', () => {
...
+    });
+
+    expect(onChange).toBeCalledTimes(1);
+    expect(onChange).toBeCalledWith(
+      expect.objectContaining({
+        bubbles: true,
+        cancelable: false,
+        currentTarget: null,
+        type: 'change',
+      }),
```
Comment (michalbundyra) [inline src/__tests__/useForm/register.test.tsx:1048]: I am not sure what is the better way to check if we are getting the correct event

File: src/types/validator.ts [hunk 2]
```diff
@@ -39,4 +39,6 @@ export type RegisterOptions<
   value: FieldPathValue<TFieldValues, TFieldName>;
   setValueAs: (value: any) => any;
   shouldUnregister?: boolean;
+  onChange?: (event: any) => void;
```
Comment (michalbundyra) [inline src/types/validator.ts:42]: wasn't sure if it should be void or maybe `Promise<boolean | void>`

File: src/useForm.ts [hunk 3]
```diff
@@ -928,8 +928,22 @@ export function useForm<
         ? ({ name: name as InternalFieldName } as UseFormRegisterReturn)
         : {
             name,
-            onChange: handleChange,
-            onBlur: handleChange,
+            onChange: options.onChange
+              ? (event: any) => {
+                  const result = handleChange(event);
```
Comment (michalbundyra) [inline src/useForm.ts:933]: per provided examples, I am calling react-hook-form handler first, and then the custom one, not sure if it makes any difference?

Maybe we can just do:
```typescript
options.onChange(event);
return handleChange(event);
```

File: src/useForm.ts [hunk 4]
```diff
@@ -928,8 +928,22 @@ export function useForm<
...
+                  options.onChange && options.onChange(event);
```
Comment (michalbundyra) [inline src/useForm.ts:934]: imo this condition is redundant as it is already inside `options.onChange` but linter and tests were complaining, that `options.onChange` might be undefined.

**Review by michalbundyra — COMMENTED** [review]
@bluebill1049 This is my attempt of adding custom handlers for onBlur and onChange. 

See my comments below and please let me know what do you think. Thanks

**Comment by bluebill1049** [comment]
Thanks for the PR, I am still keeping an eye on the feature request before the implementation itself. If you need this feature badly for your app, I suggest to fork the repo and publish your own for now. The current state of the lib is pretty stable so I wouldn't be so worried about updates.

**Comment by bluebill1049** [comment]
pump.

**Comment by cwagner22** [comment]
Any update? I spent hours migrating to v7, but now I realized I'd have to spend more time moving the register out of my components to be able to use my custom onChange callbacks. Btw see https://github.com/react-hook-form/react-hook-form/issues/5428, we also need to add `name={firstName.name}` which is missing in the doc for custom onChange

**Comment by bluebill1049** [comment]
thanks for the contribution and patient @michalbundyra will push that into the next minor update.

**Comment by bluebill1049** [comment]
@michaelbirchler FYI: got few bugs to roll out before I merge this PR since I just saw you in the issue.

**Comment by michalbundyra** [comment]
@bluebill1049 no problem, thanks 👍

**Comment by bluebill1049** [comment]
doc update: https://github.com/react-hook-form/documentation/pull/668

**Pull request merged**